include/PluginPinConnector.h

```diff
@@ -142,22 +198,539 @@ class LMMS_EXPORT PluginPinConnector
+
+		if constexpr (rc == 0b00)
+		{
+			// Both track channels bypassed - nothing to do
+			return;
+		}
+
+		// We know at this point that we are writing to at least one of the output channels
+		// rather than bypassing, so it is safe to set the output buffer of those channels
+		// to zero prior to accumulation
+
+		if constexpr (rc == 0b11)
+		{
+			// TODO: std::memcpy?
+			std::fill_n(outPtr, inOut.frames, SampleFrame{});
+		}
+		else
+		{
+			for (f_cnt_t frame = 0; frame < inOut.frames; ++frame)
+			{
+				if constexpr ((rc & 0b10) != 0)
+				{
+					outPtr[frame].leftRef() = 0.f;
+				}
+				if constexpr ((rc & 0b01) != 0)
+				{
+					outPtr[frame].rightRef() = 0.f;
+				}
+			}
+		}
+
+		// Counters for # of in channels routed to the current pair of out channels
+		ch_cnt_t numRoutedL = 0;
+		ch_cnt_t numRoutedR = 0;
+
+		for (pi_ch_t inChannel = 0; inChannel < in.channels(); ++inChannel)
+		{
+			const SampleType<layout, const SampleT>* inPtr = in.buffer(inChannel);
+
+			if constexpr (rc == 0b11)
+			{
+				// This input channel could be routed to either left, right, both, or neither output channels
+				if (m_pc->m_out.enabled(outChannel, inChannel))
+				{
+					++numRoutedL;
+					if (m_pc->m_out.enabled(outChannel + 1, inChannel))
+					{
+						++numRoutedR;
+						for (f_cnt_t frame = 0; frame < inOut.frames; ++frame)
+						{
+							outPtr[frame].leftRef() += inPtr[frame];
+							outPtr[frame].rightRef() += inPtr[frame];
+						}
+					}
+					else
+					{
+						for (f_cnt_t frame = 0; frame < inOut.frames; ++frame)
+						{
+							outPtr[frame].leftRef() += inPtr[frame];
+						}
+					}
+				}
+				else if (m_pc->m_out.enabled(outChannel + 1, inChannel))
+				{
+					++numRoutedR;
+					for (f_cnt_t frame = 0; frame < inOut.frames; ++frame)
+					{
+						outPtr[frame].rightRef() += inPtr[frame];
+					}
+				}
+			}
+			else if constexpr (rc == 0b10)
+			{
+				// This input channel may or may not be routed to the left output channel
+				if (!m_pc->m_out.enabled(outChannel, inChannel)) { continue; }
+
+				++numRoutedL;
+				for (f_cnt_t frame = 0; frame < inOut.frames; ++frame)
+				{
+					outPtr[frame].leftRef() += inPtr[frame];
+				}
+			}
+			else if constexpr (rc == 0b01)
+			{
+				// This input channel may or may not be routed to the right output channel
+				if (!m_pc->m_out.enabled(outChannel + 1, inChannel)) { continue; }
+
+				++numRoutedR;
+				for (f_cnt_t frame = 0; frame < inOut.frames; ++frame)
+				{
+					outPtr[frame].rightRef() += inPtr[frame];
+				}
+			}
+		}
+
+		// If num routed is 0 or 1, either no plugin channels were routed to the output
+		// and the output stays zeroed, or only one channel was routed and normalization is not needed
+
+		if (numRoutedL > 1)
+		{
+			if (numRoutedR > 1)
+			{
+				// Normalize output - both channels
+				for (f_cnt_t frame = 0; frame < inOut.frames; ++frame)
+				{
+					outPtr[frame].leftRef() /= numRoutedL;
```

Comment:
That's a good observation, and I've thought about it myself. The intention of the division behavior was to average the audio in order to make mono plugins sound exactly the same and have the same volume using the pin connector as it would when having two duplicated mono plugin instances (one for left channel, one for right channel), but it creates a problem for plugins with multiple outputs as you've noticed, since a drum machine with 10 outputs will have 1/10th the volume when all 10 outputs are used than when only one output is used.

I talked to @LostRobotMusic about this recently, and we concluded that the pin connector should be summing the channels together rather than averaging them. That is what REAPER's pin connector does.

However, one potentially undesired outcome of this is that mono-input/mono-output plugins connected to both L/R channels of LMMS on their input and output will effectively double the volume of audio passing through it. This is because both of LMMS's track channels are summed together going to the plugin's single input, then the plugin's single output is routed back to both LMMS track channels, so for an effect plugin that does nothing to the audio and is just a passthrough: `Core_L_out = Core_R_out = Core_L_in + Core_R_in` instead the current `Core_L_out = Core_R_out = (Core_L_in + Core_R_in) / 2`.

I could potentially add an option to choose between "averaging" and "summing" behavior rather than making it always summing. If I did that and made "averaging" the default, it would be easier to migrate Lv2 (and maybe LADSPA) away from the current practice of using 2 mono plugin instances to turn mono audio into stereo audio. Though if the "summing" method is unconditionally chosen, maybe the migration could use an Amplifier effect added to the effect chain to compensate for the audio being doubled in volume.

I plan to buy REAPER and do some experimenting to verify that my pin connector behaves identically to the REAPER pin connector it is modeled after.

Reply:
I just bought REAPER and tested it with a mono-input/mono-output plugin.

Here's what it did for the default connections:
![reaper_mono_plugin_default_pin_connector_connections](https://github.com/user-attachments/assets/818cad86-1eea-43dd-b2ae-e50004b97bb6)

It only connects the 1st track channel to the plugin input but connects the plugin output to both track channels. This is a little strange in my opinion, but it does avoid doubling the volume this way.

Interestingly, when I use the "Reset" button, it removes the connection to Track Channel 2 on the output.